plugins/pulp_python/plugins/models.py

```diff
@@ -1,120 +1,150 @@
-                        metadata_file = member
-            if not metadata_file:
-                msg = _('The archive at %(path)s does not contain a PKG-INFO file.')
-                msg = msg % {'path': archive_path}
-                raise ValueError(msg)
-
-            metadata_file_name = metadata_file.name
-            metadata_file = package_archive.extractfile(metadata_file)
-            metadata = metadata_file.read()
-
-            # Build a list of tuples of all the attributes found in the metadata. Ignore attributes
-            # with a leading underscore, as they are not part of the metadata.
-            attrs = dict()
-            try:
-                for attr in REQUIRED_ATTRS:
-                    attrs[attr] = re.search('^%s: (?P<field>.*?)\s*$' % cls._metadata_label(attr),
-                                            metadata, flags=re.MULTILINE).group('field')
-            except AttributeError:
-                msg = _('The PKG-INFO file is missing required attributes. Please ensure that the '
-                        'following attributes are all present: %(attrs)s')
-                msg = msg % {
-                    'attrs': ', '.join([cls._metadata_label(attr) for attr in REQUIRED_ATTRS])}
-                raise ValueError(msg)
-
-            # Add the filename, checksum, and checksum_type to the attrs
-            attrs['_filename'] = '%s-%s.tar%s' % (attrs['name'], attrs['version'], compression_type)
-            attrs['_checksum'] = checksum
-            attrs['_checksum_type'] = DEFAULT_CHECKSUM_TYPE
-            attrs['_metadata_file'] = metadata_file_name
-            package = cls(**attrs)
-            return package
-        finally:
-            if 'package_archive' in locals():
-                package_archive.close()
+    def from_json(cls, package_data, release, project_data):
+        """
+        Create and return (but do not save) an instance of a single Python package object from the
+        metadata available from PyPI JSON.
+
+        :param package_data: metadata specific to a file
+        :type  package_data: dict
+        :param release: version number
+        :type  release: basestring
+        :param project_data: metadata that applies to all versions
+        :type  project_data: dict
+        :return: instance of a package
+        :rtype:  pulp_python.plugins.models.Package
+        """
+        package_attrs = {}
+
+        package_attrs['version'] = release
+        package_attrs['name'] = project_data['name']
+        package_attrs['author'] = project_data['author']
+        package_attrs['summary'] = project_data['summary']
+
+        package_attrs['filename'] = package_data['filename']
+        package_attrs['path'] = package_data['path']
+        package_attrs['packagetype'] = package_data['packagetype']
+        package_attrs['md5_digest'] = package_data['md5_digest']
+
+        # If we are syncing from PyPI, there will be no `checksum`, but will be `md5_digest`
```

Comment:
Based on https://github.com/pypa/warehouse/pull/891 I _think_ we can't expect an ``md5_digest`` when warehouse is released, but we can expect one or more ``{digest_type}_digest`` fields (currently just md5 and sha256 it looks like).